### ISSUE TITLE
CRITICAL: entrenar trains 66x slower than PyTorch — val_ppl 793 vs 12 at step 2K

### ISSUE THREAD

**Issue by noahgift** [paiml/entrenar#309]
## Critical Finding

PyTorch canary (HuggingFace LlamaForCausalLM) reaches val_ppl=12.1 in 2000 steps
on the SAME config and data that entrenar needs 15K+ steps to reach val_ppl=129.

## Evidence

| Step | PyTorch | Entrenar (v15) | Ratio |
|------|---------|---------------|-------|
| 0 | loss=10.65 | loss=10.40 | 1.0x |
| 1000 | val_ppl=45.2 | val_ppl=805 | 18x |
| 2000 | val_ppl=12.1 | val_ppl=793 | 66x |

Same config: 350M LLaMA, AdamW lr=3e-4, cosine, batch=32K tokens, codeparrot-clean, seed=123.

## Canary command

```bash
PYTHONUNBUFFERED=1 uv run \
    --extra-index-url https://download.pytorch.org/whl/cu128 \
    --with 'torch,transformers,pyarrow' \
    scripts/canary_pytorch.py --steps 5000 --device cuda --eval-interval 1000 --seed 123
```

## Possible root causes

1. **Model parameter count mismatch**: canary has 432M, entrenar has 370M (different hidden→FFN sizing)
2. **Weight initialization difference**: PyTorch uses kaiming_uniform, entrenar may use different init
3. **Cross-entropy loss calculation**: possible off-by-one in sequence shift or masking
4. **Gradient computation**: a backward kernel may compute incorrect gradients (see ALB-038-044 history)
5. **AdamW implementation**: beta correction, epsilon handling, or weight decay application
6. **Data batching**: canary uses simple sequential iteration, entrenar uses shuffled + streaming
7. **Mixed precision**: canary is fp32, entrenar uses fp16 CUDA — possible precision loss

## Impact

All 16 training runs achieved suboptimal convergence. The model architecture
and data CAN produce val_ppl=12 in 2K steps (PyTorch proof). entrenar's training
produces val_ppl=805 at the same point — something fundamental is wrong.

This is the highest-priority bug in the sovereign stack.
